framework/src/main/java/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -546,15 +577,35 @@ private AnnotationMirror createAnnotationForStringConcatenation(
         List<String> leftValues = getStringValues(leftOperand, p);
         List<String> rightValues = getStringValues(rightOperand, p);
 
+        boolean allowNullStringConcat =
+                atypefactory.getContext().getChecker().hasOption("nullStringsConcatenation");
+
         if (leftValues != null && rightValues != null) {
             // Both operands have known string values, compute set of results
             List<String> concatValues = new ArrayList<>();
-            if (leftValues.isEmpty()) {
-                leftValues = Collections.singletonList("null");
+
+            if (allowNullStringConcat) {
+                if (isNullable(leftOperand)) {
+                    leftValues.add("null");
+                }
+                if (isNullable(rightOperand)) {
+                    rightValues.add("null");
+                }
             }
-            if (rightValues.isEmpty()) {
-                rightValues = Collections.singletonList("null");
+
+            if (leftOperand instanceof StringConversionNode) {
+                if (((StringConversionNode) leftOperand).getOperand().getType().getKind()
```

Comment:
Shouldn't this be in `isNullable`?

Reply:
The reason why I haven't included this in `isNullable` is because `isNullable` performs a more complex analysis about null nodes. Here it just checks that the operand is not the `null` reference. Please refer to the comment I added to Vlastimil's change request ([link](https://github.com/typetools/checker-framework/pull/2797#discussion_r335646600)).

Reply:
Right, the `null` literal should have `@StringVal("null")` even without `-AnullStringsConcatenation`.  This would be clearer if you add this check to `isNullable` and then repeat the check in an else branch like this:
```
if (nullStringConcat) {
  ...
} else {
  if (leftOperand instanceof StringConversionNode) {
    if (((StringConversionNode) leftOperand).getOperand().getType().getKind()
...
```